### DISCUSSION TITLE
How should we approach the trancript issues?

### DISCUSSION THREAD

**Discussion by KVonGit** [textadventures/quest#1321] [Developing Quest]
@alexwarren 

We are pretty much down to the transcript and the `clearScreen` function, as far as issues I introduced in 5.8.

Things got more out of hand in **playercore.js** than any other file, and most of it is completely my doing.

Easy one first, and it's also the major one, because it effects every game by default, without the player enabling anything:

### `clearScreen`
The Quest forum was **very** active from 2017 to 2018, especially the earlier part of '18. (In fact, I occasionally wish I could see the site's numbers from that time.)

There was literally always someone posting about Quest for quite a few consecutive months. Lots of code being discussed, lots of games getting beta-tested, and lots of folks asking about transcripts -- or at least a way to print all the game's text.

Angel Wedge posted some code that would modify `clearScreen()` so it would add a class to hide the text rather than dumping it. (Note that he wasn't posting it to get it added to Quest; he was simply having fun and helping someone out simultaneously.)

Among other ways to access this same text, this is one of the things that was changed in Quest 5.8.

We modified `clearScreen` as described, and included a boolean variable named `saveClearedText` to have an option to use the old `clearScreen` script. That boolean is `true` by default in Quest 5.8, and it slows big games down as the text continues to make the document larger and larger.

I had recently discovered the jQuery-UI pop-up that was already included in Quest, and we ended up making `showPopup` functions for everyone to have included by default. So, we ended up creating a `showScrollback` function that displays all the text in divOutput, removing the "hidden" class beforehand, in a dialog window.

While on the subject, these pop-up windows are larger than the screen on a mobile device, and they have been known to break walkthroughs when popping up and lacking the decency to close themselves during walkthroughs for the greater good of the game!

Speaking of walkthroughs, people (including yours truly) have occasionally wished the walk-through toggled a boolean to check from a Quest script, but I digress.

**Point:**

I think we should remove the jQuery-UI pop-up functions I added -- and the `showScrollback` function along with them.

If we do that, the mod to `clearScreen` would be pointless, and that could be reverted. I don't that anyone even knew this stuff was in the code, unless it cause them problems.

Now, this is where I can't do this without altering the transcript's code, because Quest 5.8 will display an in-game transcript in a dialog window just like the log did. And this is the same code `showScrollback` uses.

So, two roads diverge in a yellow wood:
- Submit one PR that reverts `clearScreen`, and leave the rest to remove in the transcript PR
- Submit a PR that modifies more than one thing

I'm thinking the former, but I wanted to explain this beforehand.

...and because this will be continued in the transcript question(s) I'll have for the transcript PR: should I patch what was meant to work in 5.8, then do a separate PR with the major changes after that? Or just one with the major changes)

**Comment by alexwarren** [comment]
I'm not entirely clear what's been added, and what the proposal(s) are here.

It sounds like Quest 5.8 got a new feature to show the entirety of a game transcript (including sections that were cleared when the screen was cleared). This transcript is shown in a pop-up dialog. Questions:

- how do you open this dialog? Is there a button the player can press, or does a game author trigger it via JS?
- why would you want the entire game transcript in a dialog in the first place...?

You're now suggesting removing the `showScrollback` function, so does that remove this dialog entirely?

As for how many PRs, I generally prefer one PR to do one thing. But if the "thing" is "remove the transcript functionality", then it's fine for that to be a fairly big PR, rather than trying to remove small pieces of it at a time without breaking other things.

**Reply by alexwarren** [reply]
Also side note while you're in this area - it looks like `playercore.js` got duplicated at some point (from the commit history I see [in May 2017](https://github.com/textadventures/quest/commits/f680fc098e7dfb47ea7673526c2a53284d0cd21b/WebPlayer/playercore.js)).

There should only be one copy of this, the one in the `PlayerController` directory.

So when making changes make sure the files match (assuming/hoping they do already). Ideally we should delete the duplicate one in `WebPlayer`, as that should be copied automatically by the build process. Maybe we should log a separate issue for that, but if you delete the extra one as part of your changes (assuming things still work properly afterwards) that's fine by me.

**Reply by KVonGit** [reply]
>You're now suggesting removing the `showScrollback` function, so does that remove this dialog entirely?
>
>As for how many PRs, I generally prefer one PR to do one thing. But if the "thing" is "remove the transcript functionality", then it's fine for that to be a fairly big PR, rather than trying to remove small pieces of it at a time without breaking other things.

Well, I really think Quest should have transcript functionality. I just know I didn't implement it very well. We have recently fixed all the issues with the actual transcript stuff, though.

...but that's not *really* related to the "scrollback", except they serve the same purpose (and both use a pop-up dialog window).

(The "scrollback" simply uses the data from` #divOutput`, and the transcript's data comes from `addText`.)

---
It's hard to explain, because things are a little muddled. The best approach I can think of is to try to focus on the "scrollback" functionality first, as it works differently than the transcript. The only thing that binds them is they both rely on a pop-up dialog window to display the data in-game.

---
The main thing concerning the "scrollback" is that `saveClearedText` should be `false` by default. It is currently `true`, and that seems to slow games down when there is a good amount of text, especially online. This effects every game, and the `showScrollback` function was never even fully implemented. So, no one even knows how to get any use out of this, considering they'd even want to do so in the first place.

The "scrollback" was from before we came up with the transcript functionality, you see. Pixie left it in the code as an option, though, to make Quest more customizable. Online authors can't really add JS very easily, and all that needs to be done is create a command that calls `JS.showScrollback()` to implement this.

...but I don't think he purposefully left `saveClearedText` set to `true` by default.

---
The second issue involves the pop-up dialog windows. They're really only problematic when using the mobile player, and the issue is the pop-up dialog window's dimensions. It's too wide. You can only see half of it on a mobile device, although you **can** scroll around to find the close button. I think this could be solved by changing this pop-up dialog to fullscreen settings, if we decided to leave the pop-up dialog functionality intact.

The pop-up dialog windows were used for in-game viewing of the log, the transcript, and this scrollback data. This was included so all these things could be accessed by online users.

NOTE: Pixie put a lot of effort into making things work just as well online as they do offline, but we simply didn't understand how most things work in the WebPlayer app. (The biggest example of that is DevMode, which we will eventually get around to.)

Anyway, there are three things that are still in Quest's code right now that keeps me up at night (but only one concerns the "scrollback" functionality):
`saveClearedText` is `true`, and this negatively effects every Quest game by default

---
Also note that I only dislike the pop-up dialog functionality I introduced because:
- it is not consistent across devices with different screen dimensions (which could likely be fixed, but I don't know how)
- it messes up walkthroughs (which could be avoided if Quest had an attribute to toggle while the walkthrough was running)

Someone more skilled than I am could probably fix this in less than five minutes, if desired. But we've found a way to use localStorage for transcripts now, so there's really no need to use a pop-up dialog, if we go the localStorage route online -- but I digress.

---
The best idea I have is to submit one little PR that would solve the issue related to the "scrollback", which would basically just be changing `saveClearedText` to `false` in **playercore.js**.

Then, we could move on to the transcript separately. I could do a PR for that which would fix the existing issues, and that would be the best way I could 'explain' what's going on with all that existing code. After that, a real PR with the proposed changes to the transcript.

Either way, `saveClearedText` in **playercore.js** should not be set to `true` right now. In the end, it will either be changed to false, or removed. So, that's the PR I'm about to submit. It would fix something right now, regardless of any other decisions. If it gets closed without getting merged that's cool. I just need to get it off my mind. (It haunts me.)

**Reply by KVonGit** [reply]
> Also side note while you're in this area - it looks like `playercore.js` got duplicated at some point (from the commit history I see [in May 2017](https://github.com/textadventures/quest/commits/f680fc098e7dfb47ea7673526c2a53284d0cd21b/WebPlayer/playercore.js)).
> 
> There should only be one copy of this, the one in the `PlayerController` directory.
> 
> So when making changes make sure the files match (assuming/hoping they do already). Ideally we should delete the duplicate one in `WebPlayer`, as that should be copied automatically by the build process. Maybe we should log a separate issue for that, but if you delete the extra one as part of your changes (assuming things still work properly afterwards) that's fine by me.

The one in 'PlayerController' replaces the one in 'WebPlayer' when I build the project[^1]. I just wasn't sure how it works on your end, so I've always tried to copy it and include it to be on the safe side. I [posted about it last month](https://github.com/textadventures/quest/discussions/1245), but more important things were happening so I closed the discussion for sanity's sake.

Anyway, I notice a `.gitignore` in 'WebPlayer', but it doesn't seem to work, haha!

```gitignore
playercore.*
```

I'm thinking you just need to change the main `.gitignore` on your end. I'll create an Issue real quick. 👍 👍 

[^1]: This is wonky. It definitely overwrites the one in 'WebPlayer' when I build the debug build. I think I have to clean the project first for it to do this in the Release build, but I'm still not certain it does it then. I always built Debug, then Release, then Debug, then Release again to be sure everything ends up in 'WebPlayer' and 'WebEditor'. (This was before you showed me how to Publish from VS.)
